### PR TITLE
feat: impl repetitive execution msm gpu

### PR DESCRIPTION
**Pull request by GideokKim**
# Description

This PR implements repetitive execution msm gpu.

- This modification allows calculations regardless of msm size on any GPU.
- Before modification, if the msm size was 26 or more, it was not calculated as a memory limit on RTX-4090.
- Before modification, when using lower performance GPUs, calculations are not possible due to memory limitations even with the smaller msm size.

### CODE REVIEW

File: tachyon/device/gpu/gpu_memory.cc [hunk 1]
```diff
@@ -162,4 +162,30 @@ gpuError_t GpuMemGetInfo(size_t* free, size_t* total) {
...
+  size_t available_memory = 0;
+
+  gpuError_t error = GpuMemGetInfo(&free_memory, &total_memory);
+  if (error != gpuSuccess) {
+    GPU_LOG(ERROR, error) << "Failed to GpuMemGetInfo()";
+    return available_memory;
+  }
+
+  // NOTE(GideokKim): CUDA is not guaranteed to be able to allocate all of the
+  // memory that the OS reports as free returned by cudaMemGetInfo function.
```
Comment (chokobole) [inline tachyon/device/gpu/gpu_memory.cc:176]:
```suggestion
  // memory that the OS reports as free returned by |cudaMemGetInfo| function.
```

File: tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h [hunk 2]
```diff
@@ -28,7 +28,7 @@ namespace {
 // the cub library to set some free memory. In the case of |bucket_memory_size|,
 // it uses more memory than the actual formula, so it was modified to an
 // empirically more appropriate formula. See
-// https://dev.ingonyama.com/icicle/primitives/msm
+// https://dev.ingonyama.com/icicle/primitives/msm#memory-usage-estimation
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:32]: Please squash the comments into the previous commit.

File: tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h [hunk 3]
```diff
@@ -167,14 +167,32 @@ bool IcicleMSM<bn254::G1AffinePoint>::Run(const BaseContainer& cpu_bases,
...
-  ret = ::bn254::projective_t::to_montgomery(ret);
-  *cpu_result = base::bit_cast<ProjectivePoint<Curve>>(ret);
+  size_t bitsize = static_cast<size_t>(
+      (config_->bitsize == 0) ? ::bn254::scalar_t::NBITS : config_->bitsize);
+
+  size_t divisions = DetermineMsmDivisionsForMemory(
+      sizeof(::bn254::scalar_t), sizeof(::bn254::affine_t),
+      sizeof(::bn254::projective_t), bases_size, config_->c, bitsize,
+      static_cast<size_t>(config_->precompute_factor),
+      static_cast<size_t>(config_->batch_size));
```
Comment (chokobole) [inline tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:178]: Please call this function only once during initialization! It's not a cheap operation, and in most cases, the available memory will remain similar throughout the program's execution.

Comment (GideokKim) [reply tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:178]: In that case, free memory is also checked only when initializing. Won't free memory fluctuate significantly during program execution? If there is a process that occupies a large amount of free memory, memory limitation issues may occur.

Comment (chokobole) [reply tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:178]: In production, I believe this situation rarely happens. The machine that produces a ZK proof probably will be ran with this purpose only!

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by chokobole — APPROVED** [review]
LGTM

File: tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h [hunk 4]
```diff
@@ -17,6 +20,77 @@
...
+  size_t shift = 0;
+  size_t log_msm_size = base::bits::Log2Ceiling(msm_size);
+
+  for (size_t number_of_divisions = 0; number_of_divisions < log_msm_size;
+       ++number_of_divisions) {
+    // See
+    // https://github.com/ingonyama-zk/icicle/blob/0cb0b49b/icicle/src/msm/msm.cu#L429-L431
+    size_t c = (user_c == 0) ? static_cast<size_t>(std::max(
+                                   base::bits::Log2Ceiling(msm_size) - 4, 1))
+                             : user_c;
```
Comment (batzor) [inline tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:51]: What does `c` represent here? Can we give more meaningful name?

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h (hunk 4 above)
Comment (GideokKim) [reply tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:51]: `MSMConfig` also uses `c`, so I used it as is. I think I should just add the part with an explanation of `c` as a comment. What do you think?

**Review by batzor — APPROVED** [review]
LGTM

File: tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h [hunk 5]
```diff
@@ -17,6 +20,78 @@
...
 namespace tachyon::math {
 
+namespace {
+
+// NOTE(GideokKim): The formula for memory usage estimation provided in the
+// document did not match the actual memory allocation, so some of the formula
+// was modified. |scalars_memory_size| and |points_memory_size| are exactly the
+// same, and |scalar_indices_memory_size| internally uses the sort function of
+// the cub library to set some free memory. In the case of |bucket_memory_size|,
+// it uses more memory than the actual formula, so it was modified to an
```
Comment (ashjeong) [inline tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:31]: bucket -> buckets, & change the wording slightly.
```suggestion
// the cub library to set some free memory. |buckets_memory_size| uses more memory than the actual formula, so it was modified to an...

**Review by ashjeong — APPROVED** [review]
LGTM!

**Pull request merged**